### PR TITLE
protocol: fix registerStandardSchemes api

### LINKED ISSUE #5303: When served over a custom protocol, only protocol absolute urls trigger a download when parsed
* Electron version: all
* Operating system: all

When an HTML document is served over a custom protocol, only url's starting with a protocol will trigger a request for an asset.

I.E. the following triggers a request

```html
<img src='peer://foo/bar/buzz/bazz.png'>
```

While the following do not

```html
<img src='/bar/buzz/bazz.png'>
<img src='./bar/buzz/bazz.png'>
<img src='//bar/buzz/bazz.png'>
```

This effectively breaks serving websites over custom protocols.

I created a set of unit tests that demonstrate this here: https://github.com/retrohacker/electron/blob/protocol_request/spec/api-protocol-spec.js#L108-L194

### PR DESCRIPTION
**Pull request by deepak1556**
Fixes #5303 

~~Depends on https://github.com/electron/libchromiumcontent/pull/201~~

### CODE REVIEW

File: spec/api-protocol-spec.js [hunk 1]
```diff
@@ -394,7 +396,7 @@ describe('protocol module', function () {
       var handler = function (request, callback) {
         callback(fakeFilePath)
       }
-      protocol.registerBufferProtocol(protocolName, handler, function (error) {
+      protocol.registerFileProtocol(protocolName, handler, function (error) {
```
Comment (zcbenz) [inline spec/api-protocol-spec.js]: Is there a reason behind this change?

**Comment by zcbenz** [comment]
It seems that the specs have passed even though we don't have electron/libchromiumcontent#201 yet.

File: spec/api-protocol-spec.js (hunk 1 above)
Comment (deepak1556) [inline spec/api-protocol-spec.js]: not related to the PR, the specs were inside `protocol.registerFileProtocol`, so thought they are supposed to test it ?

**Comment by deepak1556** [comment]
> It seems that the specs have passed

Yeah, I am not sure what to do. Should we patch ?

File: spec/api-protocol-spec.js (hunk 1 above)
Comment (zcbenz) [inline spec/api-protocol-spec.js]: That makes sense.

**Comment by zcbenz** [comment]
> Yeah, I am not sure what to do. Should we patch ?

The test case just always pass, so it can not test out whether `protocol.registerStandardSchemes` is working. A test case should fail when there is no fix, and pass after the fix gets merged.

**Comment by deepak1556** [comment]
Have made a separate PR to show the failing spec.

**Comment by zcbenz** [comment]
So we don't actually need electron/libchromiumcontent#201 to fix #5303?

**Comment by deepak1556** [comment]
Have made the changes, thanks! Now

* `Protocol` module can be initialized before the ready event.
* `registerStandardSchemes` can only be used before `ready` event and other methods only after the `ready` event. 

A possible problem that might arise, is when making protocol module work with partitions. If we hook protocol as a property of session, then it can be used only after ready event which will conflict with `registerStandardSchemes`. So how are we going make protocol module respect sessions ?

1. We could provide a `fromPartition` method
2. Move the `registerStandardSchemes` api to `app` module.

**Comment by zcbenz** [comment]
I prefer to use a less-intrusive way like this:

```js
const {createProtocolObject, registerStandardSchemes} = process.atomBinding('protocol')
exports.registerStandardSchemes = function (schemes) {
  if (isReady) {
    throw new Error()
  }
  registerStandardSchemes(schemes)
}

app.once('ready', function () {
  protocol = createProtocolObject();
  // Be compatible with old APIs.
  for (let method in protocol) {
    exports[method] = protocol[method].bind(protocol)
  }
})
```

So `registerStandardSchemes` is moved away from `api::Protocol` as an independent API, and the creation of protocol object is delayed in `createProtocolObject`.

After we have multi-session support for protocol, the `createProtocolObject()` will become `session.defaultSession.protocol`.

> A possible problem that might arise, is when making protocol module work with partitions. If we hook protocol as a property of session, then it can be used only after ready event which will conflict with registerStandardSchemes.

After moving `registerStandardSchemes` out of `api::Protocol`, it will be solved automatically.

**Comment by deepak1556** [comment]
Thanks for the explanation! have made the changes.

**Comment by zcbenz** [comment]
:+1:

**Pull request merged**